### PR TITLE
Add functionality to read checkpoint from legacy function

### PR DESCRIPTION
**Pull request by finsberg**
## Add a test for reading legacy functions (scalar and vector) from checkpoint file. 
I added a test for reading legacy functions from checkpoint files. This test failed initially so I implemented some changes to make it work. More specifically I needed to add some logic for reading the vector, which is usually stored under `{group}/vector_0`, but for checkpoint this is stored under `{group}/{group_0}/vector`. Also added an option to pass in the checkpoint number (here 0).

To get this to work I also needed handle the case when the shape was (N, 1), instead of (N,).

## Refactor functions for reading mesh
The only difference (as far as I can see) between reading a mesh from a pure HDF5 file and from a checkpoint file is that we use `f"{group}/geometry"` in pure HDF5, and `f"{group}/coordinates"` in checkpoint. Therefore I just add one function that first tries to read from `f"{group}/geometry"` and then if the first fails read from `f"{group}/coordinates"`. Also we need to specify the cell type in  checkpoint, so I pass this as default argument and uses that if reading the cell type fails.

### CODE REVIEW

File: src/adios4dolfinx/adios2_helpers.py [hunk 1]
```diff
@@ -185,11 +185,17 @@ def read_array(
 
     arr = io.InquireVariable(array_name)
     arr_shape = arr.Shape()
-    assert len(arr_shape) == 1
+    assert len(arr_shape) >= 1  # TODO: Should we always pick the first element?
     arr_range = compute_local_range(comm, arr_shape[0])
-    assert arr_range[0] == arr_range[0]
-    arr.SetSelection([[arr_range[0]], [arr_range[1] - arr_range[0]]])
-    vals = np.empty(arr_range[1] - arr_range[0], dtype=np.dtype(arr.Type().strip("_t")))
+    assert arr_range[0] == arr_range[0]  # FIXME: Should this be arr_shape?
```
Comment (jorgensd) [inline src/adios4dolfinx/adios2_helpers.py:190]: This doesn't test anything?:P always true?:)

File: src/adios4dolfinx/legacy_readers.py [hunk 2]
```diff
@@ -353,14 +318,37 @@ def read_mesh_from_legacy_checkpoint(
         gdim=mesh_geometry.shape[1],
     )
     domain = ufl.Mesh(element)
-
     return dolfinx.mesh.create_mesh(
         MPI.COMM_WORLD, mesh_topology, mesh_geometry, domain
     )
 
+# Function for reading global cells. Might be needed later
```
Comment (jorgensd) [inline src/adios4dolfinx/legacy_readers.py:325]: Lets put this function in an issue, with some context;)

File: src/adios4dolfinx/legacy_readers.py [hunk 3]
```diff
@@ -353,14 +318,37 @@ def read_mesh_from_legacy_checkpoint(
...
+#     return global_cells.astype(np.int64)
+
 
 def read_function_from_legacy_h5(
-    comm: MPI.Intracomm, filename: pathlib.Path, u: dolfinx.fem.Function, group: str = "mesh"
+    comm: MPI.Intracomm,
+    filename: pathlib.Path,
+    u: dolfinx.fem.Function,
+    group: str = "mesh",
+    checkpoint: Optional[int] = None,
```
Comment (jorgensd) [inline src/adios4dolfinx/legacy_readers.py:351]: Could you add a docstring? Is this the "step" or N-th instance of the checkpoint in the file?

**Review by jorgensd — CHANGES_REQUESTED** [review]
Only minor stuff! Good job

File: src/adios4dolfinx/adios2_helpers.py (hunk 1 above)
Comment (finsberg) [reply src/adios4dolfinx/adios2_helpers.py:190]: I agree, but the line was there from before I I assumed maybe you wanted to test
```
arr_range[0] == arr_shape[0]
```
or something similar?

Comment (jorgensd) [reply src/adios4dolfinx/adios2_helpers.py:190]: Probably

File: src/adios4dolfinx/legacy_readers.py (hunk 2 above)
Comment (finsberg) [reply src/adios4dolfinx/legacy_readers.py]: Removed function in https://github.com/jorgensd/adios4dolfinx/pull/22/commits/6bbe5dd885852cf87d5622d5bccaa07b1b39f4be and created issue here https://github.com/jorgensd/adios4dolfinx/issues/23

File: tests/test_legacy_readers.py [hunk 4]
```diff
@@ -68,3 +89,35 @@ def test_legacy_function():
     read_function_from_legacy_h5(mesh.comm, path, w_in, group="w")
 
     assert np.allclose(wh.x.array, w_in.x.array)
+
+
+def test_read_legacy_function_from_checkpoint():
```
[the remark this answers is not in the document]
Comment (finsberg) [reply tests/test_legacy_readers.py:94]: Fixed in https://github.com/jorgensd/adios4dolfinx/pull/22/commits/36aed4a0d9e063ae55f14999d674470dc7f7b9c6

**Review by jorgensd — APPROVED** [review]

**Pull request merged**